aiida/orm/utils/links.py

```diff
@@ -122,37 +122,59 @@ def validate_link(source, target, link_type, link_label):
     if not isinstance(source, type_source) or not isinstance(target, type_target):
         raise ValueError('cannot add a {} link from {} to {}'.format(link_type, type(source), type(target)))
 
-    # Validate the outdegree of the source node
-    outgoing = source.get_outgoing(link_type=link_type)
+    def link_triple_exists(source, target, link_type, link_label):
```

Comment:
This will redefine this function at every execution of the enclosing function - wouldn't it be better to move the definition outside the function?

Reply:
Thought it didn't, but turns out the function object is recreated at every invocation even though the code block is compiled once. But still, given that this is called often I will move it out